### PR TITLE
Add ga4gh converters and tests

### PR DESCRIPTION
**Pull request by jpdna**
Depends on https://github.com/bigdatagenomics/bdg-convert/pull/35

### CODE REVIEW

File: mango-core/pom.xml [hunk 1]
```diff
@@ -176,5 +176,30 @@
...
+    </dependency>
+    <dependency>
+    <groupId>org.bdgenomics.convert</groupId>
+    <artifactId>convert-ga4gh</artifactId>
+    <version>0.2.0-SNAPSHOT</version>
+    </dependency>
+    <dependency>
+      <groupId>org.bdgenomics.convert</groupId>
+      <artifactId>convert</artifactId>
+      <version>0.2.0-SNAPSHOT</version>
```
Comment (akmorrow13) [inline mango-core/pom.xml]: @heuermh can we get a release of bdg-convert when the ga4gh changes are in?

Comment (akmorrow13) [inline mango-core/pom.xml]: this is dependent on a new release of bdg-convert.
https://github.com/bigdatagenomics/bdg-convert/issues/36

File: mango-core/src/test/resources/small.sam [hunk 2]
```diff
@@ -0,0 +1,24 @@
+@SQ	SN:1	LN:249250621
```
Comment (akmorrow13) [inline mango-core/src/test/resources/small.sam]: why was this file changed?

File: mango-core/src/test/scala/org/bdgenomics/mango/converters/GA4GHutilSuite.scala [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    assert(json === correctJson)
+  }
+
```
Comment (akmorrow13) [inline mango-core/src/test/scala/org/bdgenomics/mango/converters/GA4GHutilSuite.scala]: Add Feature test

**Comment by akmorrow13** [comment]
Looks good! Made some comments. This PR will be dependent on a release of convert (submitted issue here: https://github.com/bigdatagenomics/bdg-convert/issues/36)

File: mango-core/src/main/scala/org/bdgenomics/mango/converters/GA4GHutil.scala [hunk 4]
```diff
@@ -0,0 +1,88 @@
+package org.bdgenomics.mango.converters
```
Comment (akmorrow13) [inline mango-core/src/main/scala/org/bdgenomics/mango/converters/GA4GHutil.scala]: license

File: mango-core/src/main/scala/org/bdgenomics/mango/converters/GA4GHutil.scala [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+import org.bdgenomics.formats.avro.Variant
+import org.bdgenomics.mango.converters
+
+import scala.collection.JavaConverters._
+import scala.collection.JavaConversions._
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+
+/**
+ * Created by paschalj on 9/8/17.
```
Comment (akmorrow13) [inline mango-core/src/main/scala/org/bdgenomics/mango/converters/GA4GHutil.scala]: remove

File: mango-core/src/main/scala/org/bdgenomics/mango/converters/GA4GHutil.scala [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    // We would prefer to run the convert as a map on the RDD, however fails currently because GA4GH type
+    // is not registered in kryo
+    /*
+    val gaVariants: Array[ga4gh.Variants.Variant] = variantContextRDD.rdd.map(a => {
+       ga4gh.Variants.Variant.newBuilder(variantConverter.convert(a.variant.variant, ConversionStringency.LENIENT, logger))
+       .addAllCalls(a.genotypes.map((g) => genotypeConverter.convert(g, ConversionStringency.LENIENT, logger)).asJava)
+       .build()
+       }).collect()
+     */
+
```
Comment (akmorrow13) [inline mango-core/src/main/scala/org/bdgenomics/mango/converters/GA4GHutil.scala]: we need featureRDDtoJson

File: mango-core/pom.xml [hunk 7]
```diff
@@ -176,5 +176,30 @@
...
+      <groupId>com.google.protobuf</groupId>
+      <artifactId>protobuf-java</artifactId>
+      <version>3.0.0-beta-3</version>
+    </dependency>
+    <dependency>
+      <groupId>com.google.protobuf</groupId>
+      <artifactId>protobuf-java-util</artifactId>
+      <version>3.0.0-beta-3</version>
+    </dependency>
+    <dependency>
```
Comment (akmorrow13) [inline mango-core/pom.xml]: put these in pom.xml and remove versions

**Comment by jpdna** [comment]
Depends on my fix to bdg-convert:
https://github.com/bigdatagenomics/convert/pull/39

**Comment by jpdna** [comment]
fyi - with my bdg-convert feature fix, mango at commit: 2d4e02bc2e674bd68fd81e309a298f310c030a14 on this PR passes tests.

**Comment by heuermh** [comment]
>  Perhaps in the bdg-convert implementation need to be adding some more checks to each of these parameters before setting them on the ga4gh.Variants.Variant.Builder

Yes, unfortunately a proper implementation of a converter will have to look like this, with null check guards around every builder setter

https://github.com/heuermh/dishevelled-bio/blob/master/convert/src/main/java/org/dishevelled/bio/convert/FeatureToGff3Record.java#L107

**Comment by jpdna** [comment]
thanks @heuermh - I'll make a PR with such guards added later tonight or tomorrow, unless I hear from @akmorrow13 that she has already done it.  

@akmorrow13 - could you send me the null pointer error trace you are getting and the commit/repo its happening on so I can check if those guards fixes what you are seeing?

**Comment by akmorrow13** [comment]
Here is the error @jpdna 
```
- create JSON from GenotypeRDD *** FAILED ***
  java.lang.NullPointerException:
  at org.bdgenomics.convert.ga4gh.BdgenomicsGenotypeToGa4ghCall.convert(BdgenomicsGenotypeToGa4ghCall.java:82)
  at org.bdgenomics.convert.ga4gh.BdgenomicsGenotypeToGa4ghCall.convert(BdgenomicsGenotypeToGa4ghCall.java:42)
  at org.bdgenomics.mango.converters.GA4GHutil$$anonfun$4$$anonfun$apply$1.apply(GA4GHutil.scala:104)
  at org.bdgenomics.mango.converters.GA4GHutil$$anonfun$4$$anonfun$apply$1.apply(GA4GHutil.scala:104)
  at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
  at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
  at scala.collection.Iterator$class.foreach(Iterator.scala:743)
  at org.apache.spark.util.collection.CompactBuffer$$anon$1.foreach(CompactBuffer.scala:115)
  at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
  at org.apache.spark.util.collection.CompactBuffer.foreach(CompactBuffer.scala:30)
  at scala.collection.TraversableLike$class.map(TraversableLike.scala:245)
  at org.apache.spark.util.collection.CompactBuffer.map(CompactBuffer.scala:30)
  at org.bdgenomics.mango.converters.GA4GHutil$$anonfun$4.apply(GA4GHutil.scala:104)
  at org.bdgenomics.mango.converters.GA4GHutil$$anonfun$4.apply(GA4GHutil.scala:102)
  at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
  at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
  at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
  at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
  at scala.collection.TraversableLike$class.map(TraversableLike.scala:245)
  at scala.collection.mutable.ArrayOps$ofRef.map(ArrayOps.scala:186)
  at org.bdgenomics.mango.converters.GA4GHutil$.genotypeRDDtoJSON(GA4GHutil.scala:102)
  at org.bdgenomics.mango.converters.GA4GHutilSuite$$anonfun$3.apply$mcV$sp(GA4GHutilSuite.scala:24)
  at org.bdgenomics.utils.misc.SparkFunSuite$$anonfun$sparkTest$1.apply$mcV$sp(SparkFunSuite.scala:102)
  at org.bdgenomics.utils.misc.SparkFunSuite$$anonfun$sparkTest$1.apply(SparkFunSuite.scala:98)
  at org.bdgenomics.utils.misc.SparkFunSuite$$anonfun$sparkTest$1.apply(SparkFunSuite.scala:98)
  at org.scalatest.Transformer$$anonfun$apply$1.apply$mcV$sp(Transformer.scala:22)
  at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
  at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
  at org.scalatest.Transformer.apply(Transformer.scala:22)
  at org.scalatest.Transformer.apply(Transformer.scala:20)
  at org.scalatest.FunSuiteLike$$anon$1.apply(FunSuiteLike.scala:166)
  at org.scalatest.Suite$class.withFixture(Suite.scala:1122)
  at org.scalatest.FunSuite.withFixture(FunSuite.scala:1555)
  at org.scalatest.FunSuiteLike$class.invokeWithFixture$1(FunSuiteLike.scala:163)
  at org.scalatest.FunSuiteLike$$anonfun$runTest$1.apply(FunSuiteLike.scala:175)
  at org.scalatest.FunSuiteLike$$anonfun$runTest$1.apply(FunSuiteLike.scala:175)
  at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
  at org.scalatest.FunSuiteLike$class.runTest(FunSuiteLike.scala:175)
  at org.bdgenomics.mango.converters.GA4GHutilSuite.org$scalatest$BeforeAndAfter$$super$runTest(GA4GHutilSuite.scala:8)
  at org.scalatest.BeforeAndAfter$class.runTest(BeforeAndAfter.scala:200)
  at org.bdgenomics.mango.converters.GA4GHutilSuite.runTest(GA4GHutilSuite.scala:8)
  at org.scalatest.FunSuiteLike$$anonfun$runTests$1.apply(FunSuiteLike.scala:208)
  at org.scalatest.FunSuiteLike$$anonfun$runTests$1.apply(FunSuiteLike.scala:208)
  at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:413)
  at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:401)
  at scala.collection.immutable.List.foreach(List.scala:383)
  at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
  at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:396)
  at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:483)
  at org.scalatest.FunSuiteLike$class.runTests(FunSuiteLike.scala:208)
  at org.scalatest.FunSuite.runTests(FunSuite.scala:1555)
  at org.scalatest.Suite$class.run(Suite.scala:1424)
  at org.scalatest.FunSuite.org$scalatest$FunSuiteLike$$super$run(FunSuite.scala:1555)
  at org.scalatest.FunSuiteLike$$anonfun$run$1.apply(FunSuiteLike.scala:212)
  at org.scalatest.FunSuiteLike$$anonfun$run$1.apply(FunSuiteLike.scala:212)
  at org.scalatest.SuperEngine.runImpl(Engine.scala:545)
  at org.scalatest.FunSuiteLike$class.run(FunSuiteLike.scala:212)
  at org.bdgenomics.mango.converters.GA4GHutilSuite.org$scalatest$BeforeAndAfter$$super$run(GA4GHutilSuite.scala:8)
  at org.scalatest.BeforeAndAfter$class.run(BeforeAndAfter.scala:241)
  at org.bdgenomics.mango.converters.GA4GHutilSuite.run(GA4GHutilSuite.scala:8)
  at org.scalatest.Suite$class.callExecuteOnSuite$1(Suite.scala:1492)
  at org.scalatest.Suite$$anonfun$runNestedSuites$1.apply(Suite.scala:1528)
  at org.scalatest.Suite$$anonfun$runNestedSuites$1.apply(Suite.scala:1526)
  at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
  at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
  at org.scalatest.Suite$class.runNestedSuites(Suite.scala:1526)
  at org.scalatest.tools.DiscoverySuite.runNestedSuites(DiscoverySuite.scala:29)
  at org.scalatest.Suite$class.run(Suite.scala:1421)
  at org.scalatest.tools.DiscoverySuite.run(DiscoverySuite.scala:29)
  at org.scalatest.tools.SuiteRunner.run(SuiteRunner.scala:55)
  at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$3.apply(Runner.scala:2563)
  at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$3.apply(Runner.scala:2557)
  at scala.collection.immutable.List.foreach(List.scala:383)
  at org.scalatest.tools.Runner$.doRunRunRunDaDoRunRun(Runner.scala:2557)
  at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1044)
  at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1043)
  at org.scalatest.tools.Runner$.withClassLoaderAndDispatchReporter(Runner.scala:2722)
  at org.scalatest.tools.Runner$.runOptionallyWithPassFailReporter(Runner.scala:1043)
  at org.scalatest.tools.Runner$.main(Runner.scala:860)
  at org.scalatest.tools.Runner.main(Runner.scala)
- create JSON from VariantRDD *** FAILED ***
  java.lang.NullPointerException:
  at com.google.protobuf.AbstractMessageLite$Builder.checkForNullValues(AbstractMessageLite.java:381)
  at com.google.protobuf.AbstractMessageLite$Builder.addAll(AbstractMessageLite.java:366)
  at ga4gh.Variants$Variant$Builder.addAllAlternateBases(Variants.java:9265)
  at org.bdgenomics.convert.ga4gh.BdgenomicsVariantToGa4ghVariant.convert(BdgenomicsVariantToGa4ghVariant.java:52)
  at org.bdgenomics.convert.ga4gh.BdgenomicsVariantToGa4ghVariant.convert(BdgenomicsVariantToGa4ghVariant.java:30)
  at org.bdgenomics.mango.converters.GA4GHutil$$anonfun$3.apply(GA4GHutil.scala:87)
  at org.bdgenomics.mango.converters.GA4GHutil$$anonfun$3.apply(GA4GHutil.scala:86)
  at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
  at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
  at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
  at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
  at scala.collection.TraversableLike$class.map(TraversableLike.scala:245)
  at scala.collection.mutable.ArrayOps$ofRef.map(ArrayOps.scala:186)
  at org.bdgenomics.mango.converters.GA4GHutil$.variantRDDtoJSON(GA4GHutil.scala:86)
  at org.bdgenomics.mango.converters.GA4GHutilSuite$$anonfun$4.apply$mcV$sp(GA4GHutilSuite.scala:32)
  at org.bdgenomics.utils.misc.SparkFunSuite$$anonfun$sparkTest$1.apply$mcV$sp(SparkFunSuite.scala:102)
  at org.bdgenomics.utils.misc.SparkFunSuite$$anonfun$sparkTest$1.apply(SparkFunSuite.scala:98)
  at org.bdgenomics.utils.misc.SparkFunSuite$$anonfun$sparkTest$1.apply(SparkFunSuite.scala:98)
  at org.scalatest.Transformer$$anonfun$apply$1.apply$mcV$sp(Transformer.scala:22)
  at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
  at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
  at org.scalatest.Transformer.apply(Transformer.scala:22)
  at org.scalatest.Transformer.apply(Transformer.scala:20)
  at org.scalatest.FunSuiteLike$$anon$1.apply(FunSuiteLike.scala:166)
  at org.scalatest.Suite$class.withFixture(Suite.scala:1122)
  at org.scalatest.FunSuite.withFixture(FunSuite.scala:1555)
  at org.scalatest.FunSuiteLike$class.invokeWithFixture$1(FunSuiteLike.scala:163)
  at org.scalatest.FunSuiteLike$$anonfun$runTest$1.apply(FunSuiteLike.scala:175)
  at org.scalatest.FunSuiteLike$$anonfun$runTest$1.apply(FunSuiteLike.scala:175)
  at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
  at org.scalatest.FunSuiteLike$class.runTest(FunSuiteLike.scala:175)
  at org.bdgenomics.mango.converters.GA4GHutilSuite.org$scalatest$BeforeAndAfter$$super$runTest(GA4GHutilSuite.scala:8)
  at org.scalatest.BeforeAndAfter$class.runTest(BeforeAndAfter.scala:200)
  at org.bdgenomics.mango.converters.GA4GHutilSuite.runTest(GA4GHutilSuite.scala:8)
  at org.scalatest.FunSuiteLike$$anonfun$runTests$1.apply(FunSuiteLike.scala:208)
  at org.scalatest.FunSuiteLike$$anonfun$runTests$1.apply(FunSuiteLike.scala:208)
  at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:413)
  at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:401)
  at scala.collection.immutable.List.foreach(List.scala:383)
  at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
  at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:396)
  at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:483)
  at org.scalatest.FunSuiteLike$class.runTests(FunSuiteLike.scala:208)
  at org.scalatest.FunSuite.runTests(FunSuite.scala:1555)
  at org.scalatest.Suite$class.run(Suite.scala:1424)
  at org.scalatest.FunSuite.org$scalatest$FunSuiteLike$$super$run(FunSuite.scala:1555)
  at org.scalatest.FunSuiteLike$$anonfun$run$1.apply(FunSuiteLike.scala:212)
  at org.scalatest.FunSuiteLike$$anonfun$run$1.apply(FunSuiteLike.scala:212)
  at org.scalatest.SuperEngine.runImpl(Engine.scala:545)
  at org.scalatest.FunSuiteLike$class.run(FunSuiteLike.scala:212)
  at org.bdgenomics.mango.converters.GA4GHutilSuite.org$scalatest$BeforeAndAfter$$super$run(GA4GHutilSuite.scala:8)
  at org.scalatest.BeforeAndAfter$class.run(BeforeAndAfter.scala:241)
  at org.bdgenomics.mango.converters.GA4GHutilSuite.run(GA4GHutilSuite.scala:8)
  at org.scalatest.Suite$class.callExecuteOnSuite$1(Suite.scala:1492)
  at org.scalatest.Suite$$anonfun$runNestedSuites$1.apply(Suite.scala:1528)
  at org.scalatest.Suite$$anonfun$runNestedSuites$1.apply(Suite.scala:1526)
  at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
  at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
  at org.scalatest.Suite$class.runNestedSuites(Suite.scala:1526)
  at org.scalatest.tools.DiscoverySuite.runNestedSuites(DiscoverySuite.scala:29)
  at org.scalatest.Suite$class.run(Suite.scala:1421)
  at org.scalatest.tools.DiscoverySuite.run(DiscoverySuite.scala:29)
  at org.scalatest.tools.SuiteRunner.run(SuiteRunner.scala:55)
  at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$3.apply(Runner.scala:2563)
  at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$3.apply(Runner.scala:2557)
  at scala.collection.immutable.List.foreach(List.scala:383)
  at org.scalatest.tools.Runner$.doRunRunRunDaDoRunRun(Runner.scala:2557)
  at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1044)
  at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1043)
  at org.scalatest.tools.Runner$.withClassLoaderAndDispatchReporter(Runner.scala:2722)
  at org.scalatest.tools.Runner$.runOptionallyWithPassFailReporter(Runner.scala:1043)
  at org.scalatest.tools.Runner$.main(Runner.scala:860)
  at org.scalatest.tools.Runner.main(Runner.scala)

```

**Comment by jpdna** [comment]
> Here is the error @jpdna

and what commit/repo was this run from?

**Comment by jpdna** [comment]
@akmorrow13 - tests pass locally at commit 
109a585
when I use locally build `convert` seen at 
https://github.com/bigdatagenomics/convert/pull/40
commit ca14287

Also, I added a `GA4GHutil.variantRDDtoJSON` without genotypes as you mentioned a need for.

Let me know if you are still seeing any null pointer problems.

**Comment by akmorrow13** [comment]
Great @jpdna ! let me try this out. Do you know when a release of convert will be available?

**Comment by akmorrow13** [comment]
@jpdna can we get this finished up today? The 0.3.0-SNAPSHOT of bdg-convert is up so we are not blocked.

File: mango-core/src/main/scala/org/bdgenomics/mango/converters/GA4GHutil.scala [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    val gaReads: Array[ReadAlignment] = alignmentRecordRDD.rdd.collect.map(a => alignmentConverter.convert(a, ConversionStringency.LENIENT, logger))
+
+    // We would prefer to run the convert as a map on the RDD, however fails currently because GA4GH type
+    // is not registered in kryo
+    /*
+    val gaReads: Array[ReadAlignment] = alignmentRecordRDD.rdd.map(a => alignmentConverter
+      .convert(a, ConversionStringency.LENIENT, logger))
+      .collect()
+    */
```
Comment (akmorrow13) [inline mango-core/src/main/scala/org/bdgenomics/mango/converters/GA4GHutil.scala]: we can remove for now and stay with this
